lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp

```diff
@@ -24,132 +28,323 @@ uint64_t getOpOutputLayoutUsage(
+            }
+          }
+        }
+
+        // Greedily find the optimal configuration.
+        //
+        Config optimalConfig = getOptimalConfig(op, opsL1Usage);
+        for (const auto &[op, layout] : optimalConfig.layouts) {
+          OpMemSpecMap[op].layout = layout;
         }
 
-        currentOp = nullptr;
-        if (!l1ChainConfigs->back().isEmpty()) {
-          l1ChainConfigs->back().build();
-          l1ChainConfigs->push_back(L1ChainConfig());
+        // Override op's precedence.
+        //
+        opsPrecedence.insert(opsPrecedence.end(),
+                             optimalConfig.precedence.begin(),
+                             optimalConfig.precedence.end());
+        precedenceMap[op] = std::move(opsPrecedence);
+
+        // Update op's requiredL1Usage if the op is analyzable.
+        //
+        if (isAnalyzable(op)) {
+          uint64_t intermediateRequiredL1Usage = 0;
+          uint64_t intermediateL1Usage = 0;
+          for (auto operand : op->getOperands()) {
+            // Skip block arguments (%arg0, %arg1, ...)
+            //
+            if (::llvm::isa<mlir::BlockArgument>(operand)) {
+              continue;
+            }
+
+            Operation *operandOp = operand.getDefiningOp();
+
+            // Skip non-analyzable operands.
+            //
+            if (isAnalyzable(operandOp)) {
+              intermediateRequiredL1Usage =
+                  std::max(intermediateRequiredL1Usage,
+                           intermediateL1Usage +
+                               OpMemSpecMap[operandOp].requiredL1Usage);
+              intermediateL1Usage += getOpOutputL1Usage(
+                  operandOp, OpMemSpecMap[operandOp].layout, deviceAttr);
+            }
+          }
+          OpMemSpecMap[op].requiredL1Usage = std::max(
+              intermediateRequiredL1Usage,
+              intermediateL1Usage +
+                  getOpOutputL1Usage(op, OpMemSpecMap[op].layout, deviceAttr));
         }
       }
     }
 
-    if (l1ChainConfigs->back().isEmpty()) {
-      l1ChainConfigs->pop_back();
-    }
+    // Construct the schedule.
+    //
+    constructSchedule(func);
 
-    // Schedule
+    // Build, Resolve and Complete the L1 chain.
+    // This implementation is only here unitl we are able to merge
+    // L1ChainConfigs.
+    // TODO(fbajraktari): Fix this hack.
     //
-    (*schedule)[func] = scheduler.getSchedule();
+    l1ChainConfigs->push_back(L1ChainConfig());
+    llvm::DenseMap<Operation *, TTNNLayoutAttr> selectedOpLayout;
+    for (auto &OpMemSpec : OpMemSpecMap) {
+      OpL1MemSpec opL1MemSpec;
+      opL1MemSpec.op = OpMemSpec.first;
+      opL1MemSpec.tensorSplitFactor = 1;
+      selectedOpLayout[OpMemSpec.first] = OpMemSpec.second.layout;
+      l1ChainConfigs->back().addOpL1MemSpec(std::move(opL1MemSpec));
+    }
+    l1ChainConfigs->back().build();
+    l1ChainConfigs->back().resolve();
+    std::unordered_set<Edge> memReconfigEdges;
```

Comment:
It depends on the op and there may be a scenario where we need reconfigEdges. For example, if we greedily decide that the output of an op's operand should be in DRAM and there is enough available L1 memory this operand then we can insert a ToLayout op on that edge so we can gain on the performance. I assume this may be true depending on the op.